### PR TITLE
[OpenMPOpt] Fix incorrect end-of-kernel barrier removal

### PR DESCRIPTION
**Pull request by dwoodwor-intel**
Barrier removal in OpenMPOpt normally removes barriers by proving that they are redundant with barriers preceding them. However, it can't do this with the "pseudo-barrier" at the end of kernels because that can't be removed. Instead, it removes the barriers preceding the end of the kernel which that end-of-kernel barrier is redundant with. However, these barriers aren't always redundant with the end-of-kernel barrier when loops are involved, and removing them can lead to incorrect results in compiled code.

This change fixes this by requiring that these pre-end-of-kernel barriers also have the kernel end as a unique successor before removing them. It also changes the initialization of `ExitED` for kernels since the kernel end is not an aligned barrier.

### CODE REVIEW

**Review by jdoerfert — COMMENTED** [review]
I need to look at it in more detail, but the reasoning makes sense (as already tried to fix in https://github.com/llvm/llvm-project/commit/24656e995ae9c5893fa32db62863e55c3c70c5e2).

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2691]: Shouldn't we just delete 2729-2731 and this conditional? I think the idea that the kernel end is an aligned barrier was fundamentally flawed.

Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2691]: I'm still wrapping my head around this pass, but from my understanding it doesn't seem unreasonable to treat the kernel end as an aligned barrier. Even if the kernel exit instructions themselves are divergent and different ones are reached in different threads, it seems like we can still model the control flow as re-converging after the different kernel exits before our implicit post-kernel barrier.

My understanding is that it's legal and useful to remove barriers near the end of kernels in cases like this:

```
define void @simple_end_barrier() "kernel" {
  call void @unknown()
  call void @aligned_barrier() ; Barrier A
  ret void
}
```

It seems like we need a concept of a post-kernel barrier to do this, though, since otherwise `Barrier A` isn't redundant and can't be removed.

Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2691]: We might want to treat the kernel end special, but it is not an aligned barrier. Threads are allowed to run into the kernel end independently, and the others can do something else in the meantime, e.g., wait in an aligned barrier. What I'm trying to say is that our current abstraction is too coarse-grained. That said, you might be right that for the purpose of barrier elimination, we can continue to "handle" the kernel end.

That said, I'm still not sure if this is the right way to fix this. I think the change in multiple_blocks_functions_kernel_effects_0 actually exposes a different error we should first address.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2683]: > only be removed if they also only reach the kernel end.

That is not true, nor does it reflect what happens below. The kernel end is not an aligned barrier which is why your test case starts to work (the explicit aligned barrier does not only reach aligned barriers but also the kernel end and is consequently not removed).

Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2683]: This should be "if they also only reach the kernel end or other aligned barriers", right?

And since `IsReachingAlignedBarrierOnly` doesn't take side-effects like the `@unknown()` call into account, I'd need to check something else to make sure there aren't any extra side-effect instructions after this barrier that aren't ahead of the kernel end. Would I need to do another backwards propagation for that? It seems like `EncounteredNonLocalSideEffect` is only for side-effects preceding the execution domain.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2691]: First, can we have a few more tests though (and pre-commit the tests please). 
```
│ define void @loop_barrier2() "kernel" {
│ entry:
│   br label %loop
│
│ loop:
│   %i = phi i32 [ 0, %entry ], [ %i.next, %loop ]
│   call void @unknown()
│   call void @aligned_barrier()
│   %i.next = add nuw nsw i32 %i, 1
│   %cond = icmp ne i32 %i.next, 128
│   br i1 %cond, label %loop, label %exit
│
│ exit:
│   call void @aligned_barrier()
│   call void @aligned_barrier()
│   call void @aligned_barrier()
│   call void @aligned_barrier()
│   ret void
│ }
│
```
Also add an unknown call in between the 4 aligned barriers.
Then, replace `unknown` with a store to a generic global.
Now the fix you proposed fails and we race on that global, at least if we put an explicit barrier at the end, e.g.,:
```
│ define void @loop_barrier3() "kernel" {

│ entry:
│   br label %loop
│
│ loop:
│   %i = phi i32 [ 0, %entry ], [ %i.next, %loop ]
│   store i32 %i, ptr @G1
│   call void @aligned_barrier()
│   %i.next = add nuw nsw i32 %i, 1
│   %cond = icmp ne i32 %i.next, 128
│   br i1 %cond, label %loop, label %exit
│
│ exit:
│   call void @aligned_barrier()
│   ret void
│ }
```

---

I'm still thinking about this.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2683]: > This should be "if they also only reach the kernel end or other aligned barriers", right?

I don't think that is accurate at least not with the rest in place. See my example (loop_barrier3) with an explicit barrier in the end. The one in the loop will hit only aligned barriers. The one in the end is useless and removed. The end -> barrier elimination will skip it and eliminate the one in the loop.

Here are my current thoughts:
The end -> barrier case is different because we delete the "former"(/only) barrier, not the latter one.
The logic we have is basically trying to determine when a barrier is not actually synchronizing/communicating anything, and it is wrong in the kernel end case because the barrier can (evidently) synchronize with itself.

The logic we use right now is fundamentally flawed since end -> barrier does not say anything about other parts from barrier to somewhere else. In fact, there is no need for the threads reaching the barrier to go (directly) to the end at all. I think this is the crucial point. Our logic only holds if the barrier we are about to eliminate, due to the kernel end case, cannot reach any other synchronization point. If it can, the path from it to the end is not enough reason to remove it. Even if it can only reach other aligned barriers not, e.g.,:
```
if (tid == 7)
  G = 1;
aligned_barrier();        // currently eliminated, reached by kernel end
if (runtime_false())
  return; // kernel exit
aligned_barrier();        // currently eliminated, no side effect since the last barrier
use(G);
```

I see two paths forward:
1) Remove the kernel end special case (and change the kernel end everywhere to not be an aligned barrier). This is simple and prevents the problem, but might cause some leftover barriers.
2) Verify that each barrier we remove in the kernel end case will not reach any other sync point. The easiest solution is to start at the kernel end, and only remove barriers in blocks for which the kernel end block is the unique successor.

WDYT?

Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2683]: I also think what we're doing now with special-casing the kernel end is not the right approach, but I think there is some value in generalizing the "former" barrier removal if we can get the conditions for it right. That would help in cases like this:

```
define void @only_first_barrier_redundant(i1 %c) "kernel" {
  call void @unknown()
  br i1 %c, label %if, label %end

if:
  call void @unknown()
  call void @aligned_barrier() ; Barrier A
  br label %end

end:
  call void @aligned_barrier() ; Barrier B
  call void @unknown()
  ret void
}
```

We can't remove `Barrier B` in this case and we can't remove `Barrier A` using our normal "latter" barrier removal because it has an `unknown` call before it. But `Barrier A` is still redundant here because it's always immediately followed by `Barrier B`, so if we had "former" barrier removal in cases other than the kernel end case we could safely remove it.

As in your example with a conditional return between the barriers, one of the big challenges here is that if we're removing both former and latter barriers in the common straight-line case we'll remove the former barrier because it's immediately followed by a barrier and also the latter barrier because it's immediately preceded by a barrier. This is a stale analysis problem, though: once we remove one of the barriers it's no longer the case that the remaining barrier does not have any side effects (or only reaches aligned barriers) before/after it. We could fix this by making sure the analysis is up-to-date, either by propagating the results from barriers as we remove them or by re-computing the analysis between the first sweep removing latter barriers and the second sweep removing former barriers.

This is a more ambitious strategy that will take more time to get right, though. For now I think just doing what we are now but only removing barriers in the end -> barrier case when the kernel end is the unique successor is a reasonable solution since it should fix all of the cases where barriers are being incorrectly removed while still removing barriers in simple cases. I'll update this PR to implement this instead.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2702]: I've added the requested testcases in my update.

With this new updated the three barriers in `multiple_blocks_functions_kernel_effects_0` are back to being removed, but is this the correct behavior? It seems to me that we should preserve at least one of the barriers between the `barrier_then_write0` calls and the `write_then_barrier0` call.

Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2702]: I think you are right.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp]: I agree that this is reasonable. Also, change the initialization of ExitED, the Kernel end is never an aligned barrier.

Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp]: Done

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2702]: I'm seeing the `aligned_barrier`s from both `write_then_barrier0` and also `m3` in `multiple_blocks_functions_kernel_effects_0` added to the `multiple_blocks_functions_kernel_effects_0` end-of-function ED's `AlignedBarriers`. The `write_then_barrier0` `aligned_barrier` is ignored because it's in another function, but it looks like we skip over the `m3` `aligned_barrier` because it's already been removed and (incorrectly) remove the three other barriers because those ones precede the `m3` `aligned_barrier`. Is this the expected behavior, or should the end-of-function ED for `multiple_blocks_functions_kernel_effects_0` only have the `write_then_barrier0` `aligned_barrier` in `AlignedBarriers` and not the one in `m3`?

Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2702]: @jdoerfert, is this issue with these barriers still being removed what's holding up this change? It looks to me like the fix for this should be separate from the fix I've implemented here to only remove barriers with a straight line to the kernel end, but is there something I'm missing?

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (jdoerfert) [inline llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2632]: It is likely not to matter here but we usually avoid recursion and go with loops.
Also, const BB * const is a long way to spell auto.

**Review by jdoerfert — APPROVED** [review]
Thanks for this fix. Apologies it took me so long to review it.

File: llvm/lib/Transforms/IPO/OpenMPOpt.cpp
Comment (dwoodwor-intel) [reply llvm/lib/Transforms/IPO/OpenMPOpt.cpp:2632]: Yes, this is tail recursion so it should be functionally equivalent

**Comment by dwoodwor-intel** [comment]
I don't have write access; @jdoerfert, can you merge this on my behalf?

**Pull request merged**
